clippy_lints/src/field_reassign_with_default.rs

```diff
@@ -0,0 +1,208 @@
+
+                    // also set first instance of error for help message
+                    if first_assign.is_none() {
+                        first_assign = Some(consequtive_statement);
+                    }
+                }
+                // interrupt also if no field was assigned, since we only want to look at consequtive statements
+                else {
+                    break;
+                }
+            }
+
+            // if there are incorrectly assigned fields, do a span_lint_and_note to suggest
+            // construction using `Ty { fields, ..Default::default() }`
+            if !assigned_fields.is_empty() {
+                // take the original assignment as span
+                let stmt = &block.stmts[stmt_idx];
+
+                if let StmtKind::Local(preceding_local) = &stmt.kind {
+                    // if all fields of the struct are not assigned, add `.. Default::default()` to the suggestion.
+                    let ext_with_default = !fields_of_type(&binding_type)
+                        .iter()
+                        .all(|field| assigned_fields.contains_key(&field.name));
+
+                    let field_list = assigned_fields
+                        .into_iter()
+                        .map(|(field, value)| format!("{}: {}", field, value))
+                        .collect::<Vec<String>>()
+                        .join(", ");
+
+                    let sugg = if ext_with_default {
+                        format!("{} {{ {}, ..Default::default() }}", binding_type, field_list)
+                    } else {
+                        format!("{} {{ {} }}", binding_type, field_list)
+                    };
+
+                    // span lint once per statement that binds default
+                    span_lint_and_note(
+                        cx,
+                        FIELD_REASSIGN_WITH_DEFAULT,
+                        first_assign.unwrap_or_else(|| unreachable!()).span,
+                        "field assignment outside of initializer for an instance created with Default::default()",
+                        Some(preceding_local.span),
+                        &format!("consider initializing the variable with `{}`", sugg),
+                    );
+                }
+            }
+        }
+    }
+}
+
+/// Returns the block indices, identifiers and types of bindings set as `Default::default()`, except
+/// for when the pattern type is a tuple.
+fn enumerate_bindings_using_default<'cx, 'hir>(
+    cx: &LateContext<'cx>,
+    block: &Block<'hir>,
+) -> Vec<(usize, Symbol, &'cx TyS<'cx>)> {
+    block
+        .stmts
+        .iter()
+        .enumerate()
+        .filter_map(|(idx, stmt)| {
+            if_chain! {
+                // only take `let ...` statements
+                if let StmtKind::Local(ref local) = stmt.kind;
+                // only take bindings to identifiers
+                if let PatKind::Binding(_, _, ident, _) = local.pat.kind;
```

Comment:
```suggestion
                if let PatKind::Binding(_, _, ident, _) = local.pat.kind;
                // that are not tuples
                let ty = cx.typeck_results().pat_ty(local.pat);
                if !matches!(ty.kind(), ty::Tuple(_));
```
We can reuse the `if_chain` for this check